File: packages/core/src/client/sse.ts
export type SSEEvent = {
	data: string
	event?: string
	id?: string
	retry?: number
}

function parseBlock(block: string): SSEEvent | undefined {
	const lines = block.split(/\r\n|\r|\n/)
	let isComment = true
	let data: string | undefined
	let event: string | undefined
	let id: string | undefined
	let retry: number | undefined

	for (const line of lines) {
		if (line.startsWith(":")) continue
		isComment = false

		const colonIdx = line.indexOf(":")
		let field: string
		let val: string
		if (colonIdx === -1) {
			field = line
			val = ""
		} else {
			field = line.slice(0, colonIdx)
			val = line.slice(colonIdx + 1)
			if (val.startsWith(" ")) val = val.slice(1)
		}

		switch (field) {
			case "data":
				data = data === undefined ? val : `${data}\n${val}`
				break
			case "event":
				event = val
				break
			case "id":
				/* per spec: id containing null byte must be ignored */
				if (!val.includes("\0")) id = val
				break
			case "retry": {
				const n = Number(val)
				if (Number.isFinite(n) && n >= 0) retry = n
				break
			}
		}
	}

	if (isComment || data === undefined) return undefined

	const sse: SSEEvent = { data }
	if (event !== undefined) sse.event = event
	if (id !== undefined) sse.id = id
	if (retry !== undefined) sse.retry = retry
	return sse
}

const DEFAULT_MAX_BUFFER = 1024 * 1024

export async function* parseSSEStream(
	stream: ReadableStream<Uint8Array>,
	opts?: { maxBufferSize?: number },
): AsyncGenerator<SSEEvent, void, undefined> {
	const decoder = new TextDecoder()
	const reader = stream.getReader()
	let buffer = ""
	const maxBuffer = opts?.maxBufferSize ?? DEFAULT_MAX_BUFFER

	try {
		while (true) {
			const { done, value } = await reader.read()
			if (done) break
			buffer += decoder.decode(value, { stream: true })

			if (buffer.length > maxBuffer) {
				throw new Error(`SSE buffer exceeded ${maxBuffer} characters`)
			}

			/* split on double-newline (any combo of \r\n, \r, \n) */
			const blocks = buffer.split(/\r\n\r\n|\r\n\r|\r\n\n|\r\r\n|\n\r\n|\n\r|\r\r|\n\n/)
			/* last element is incomplete — keep in buffer */
			buffer = blocks.pop() ?? ""

			for (const block of blocks) {
				if (block.trim() === "") continue
				const event = parseBlock(block)
				if (event) yield event
			}
		}

		/* flush remaining buffer */
		if (buffer.trim() !== "") {
			const event = parseBlock(buffer)
			if (event) yield event
		}
	} finally {
		await reader.cancel()
		reader.releaseLock()
	}
}

File: packages/core/src/client-python/_sse.py
from __future__ import annotations

import re
from typing import AsyncIterator, TypedDict

import httpx


class SSEEvent(TypedDict, total=False):
    data: str
    event: str
    id: str
    retry: int


def _parse_sse_block(block: str) -> SSEEvent | None:
    lines = block.splitlines()
    is_comment = True
    data: str | None = None
    event: str | None = None
    id_val: str | None = None
    retry: int | None = None

    for line in lines:
        if line.startswith(":"):
            continue
        is_comment = False
        colon_idx = line.find(":")
        if colon_idx == -1:
            field_name = line
            val = ""
        else:
            field_name = line[:colon_idx]
            val = line[colon_idx + 1:]
            if val.startswith(" "):
                val = val[1:]

        if field_name == "data":
            data = val if data is None else f"{data}\n{val}"
        elif field_name == "event":
            event = val
        elif field_name == "id":
            # per SSE spec: id containing null byte must be ignored
            if "\0" not in val:
                id_val = val
        elif field_name == "retry":
            try:
                n = int(val)
                if n >= 0:
                    retry = n
            except ValueError:
                pass

    if is_comment or data is None:
        return None

    result: SSEEvent = {"data": data}
    if event is not None:
        result["event"] = event
    if id_val is not None:
        result["id"] = id_val
    if retry is not None:
        result["retry"] = retry
    return result


DEFAULT_MAX_BUFFER = 1024 * 1024

_SSE_BLOCK_RE = re.compile(r"\r\n\r\n|\r\n\r|\r\n\n|\r\r\n|\n\r\n|\n\r|\r\r|\n\n")


async def parse_sse_stream(
    response: httpx.Response,
    max_buffer_size: int = DEFAULT_MAX_BUFFER,
) -> AsyncIterator[SSEEvent]:
    buffer = ""
    async for chunk in response.aiter_text():
        buffer += chunk
        if len(buffer) > max_buffer_size:
            raise RuntimeError(f"SSE buffer exceeded {max_buffer_size} characters")

        # split on double-newline (any combo of \r\n, \r, \n)
        blocks = _SSE_BLOCK_RE.split(buffer)
        buffer = blocks.pop()

        for block in blocks:
            if not block.strip():
                continue
            event = _parse_sse_block(block)
            if event is not None:
                yield event

    if buffer.strip():
        event = _parse_sse_block(buffer)
        if event is not None:
            yield event


__all__ = ["SSEEvent", "parse_sse_stream"]
